### PR TITLE
[CARBONDATA-1815][PreAgg] Add AtomicRunnableCommand abstraction

### PR DESCRIPTION
**Pull request by jackylk**
Some CarbonData command like CreatePreAggregateTableCommand need to process both metadata and data which should be in an atomic fashion. These commands need to support undo if any failure. So this PR adds AtomicRunnableCommand to enforce this behavior.

 - [X] Any interfaces changed?
 No
 - [X] Any backward compatibility impacted?
 No
 - [X] Document update required?
No
 - [X] Testing done
Retest all testcases
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1486/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/execution/command/CarbonHiveCommands.scala [hunk 1]
```diff
@@ -51,7 +51,7 @@ case class CarbonDropDatabaseCommand(command: DropDatabaseCommand)
     // DropHiveDB command will fail if cascade is false and one or more table exists in database
     if (command.cascade && tablesInDB != null) {
       tablesInDB.foreach { tableName =>
-        CarbonDropTableCommand(true, tableName.database, tableName.table).run(sparkSession)
+          CarbonDropTableCommand(true, tableName.database, tableName.table).run(sparkSession)
```
Comment (QiangCai) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/execution/command/CarbonHiveCommands.scala]: not required

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 2]
```diff
@@ -150,7 +149,7 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
     DROP ~> DATAMAP ~> opt(IF ~> EXISTS) ~ ident ~ (ON ~ TABLE) ~
     (ident <~ ".").? ~ ident <~ opt(";")  ^^ {
       case ifexists ~ dmname ~ ontable ~ dbName ~ tableName =>
-        CarbonDropDataMapCommand(dmname, ifexists.isDefined, dbName, tableName)
+      CarbonDropDataMapCommand(dmname, ifexists.isDefined, dbName, tableName)
```
Comment (QiangCai) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: not required

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1493/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1504/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1508/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1910/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1510/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/389/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1519/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1921/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1528/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1549/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1948/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1557/

**Comment by QiangCai** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1957/

**Pull request closed without merging**